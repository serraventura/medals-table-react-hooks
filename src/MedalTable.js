import React from "react";
import {
  Row,
  Column,
  ColumnPoints,
  ColumnTotalPoints
} from "./styledComponents";

const MedalTable = ({ list = [], onAddGold, onAddSilver, onAddBronze }) => (
  <div>
    <Row>
      <ColumnPoints isFixed />
      <Column isFixed />
      <ColumnPoints isFixed>GOLD</ColumnPoints>
      <ColumnPoints isFixed>SILVER</ColumnPoints>
      <ColumnPoints isFixed>BRONZE</ColumnPoints>
      <ColumnTotalPoints isFixed />
    </Row>
    {list.map((i, c) => (
      <Row key={i.country}>
        <ColumnPoints counter={c}>{c + 1}</ColumnPoints>
        <Column counter={c}>
          {i.country} ({i.totalPoints})
        </Column>
        <ColumnPoints
          counter={c}
          data-test="gold"
          isClickable
          onClick={() => onAddGold(i.country, "add-gold")}
        >
          {i.gold}
        </ColumnPoints>
        <ColumnPoints
          counter={c}
          data-test="silver"
          isClickable
          onClick={() => onAddSilver(i.country, "add-silver")}
        >
          {i.silver}
        </ColumnPoints>
        <ColumnPoints
          counter={c}
          data-test="bronze"
          isClickable
          onClick={() => onAddBronze(i.country, "add-bronze")}
        >
          {i.bronze}
        </ColumnPoints>
        <ColumnTotalPoints counter={c}>{i.totalMedals}</ColumnTotalPoints>
      </Row>
    ))}
  </div>
);

export default MedalTable;
