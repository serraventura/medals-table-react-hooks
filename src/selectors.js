const selectAllCountries = state => {
  return Object.keys(state)
    .map(i => ({ country: i, ...state[i] }))
    .sort((a, b) => b.totalPoints - a.totalPoints);
};

export { selectAllCountries };
