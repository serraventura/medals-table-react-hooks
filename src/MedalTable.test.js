import React from "react";
import { shallow, mount, configure } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import MedalTable from "./MedalTable";
import { selectAllCountries } from "./selectors";

configure({ adapter: new Adapter() });

let onAddGold, onAddSilver, onAddBronze;

describe("<MedalTable />", () => {
  const listData = {
    UK: { gold: 0, silver: 0, bronze: 0, totalMedals: 0, totalPoints: 0 }
  };

  beforeEach(() => {
    onAddGold = jest.fn();
    onAddSilver = jest.fn();
    onAddBronze = jest.fn();
  });

  it("should render without crashing", () => {
    const wrapper = shallow(<MedalTable />);
    expect(wrapper).toBeTruthy();
  });

  it("should match snapshot", () => {
    const wrapper = mount(<MedalTable />);
    expect(wrapper).toMatchSnapshot();
  });

  it("should match snapshot with data", () => {
    const wrapper = mount(<MedalTable list={selectAllCountries(listData)} />);
    expect(wrapper).toMatchSnapshot();
  });

  it("should handle onAddGold prop", () => {
    const wrapper = shallow(
      <MedalTable list={selectAllCountries(listData)} onAddGold={onAddGold} />
    );

    wrapper.find('*[data-test="gold"]').simulate("click");

    expect(onAddGold).toHaveBeenCalled();
    expect(onAddGold).toHaveBeenCalledWith("UK", "add-gold");
  });

  it("should handle onAddSilver prop", () => {
    const wrapper = shallow(
      <MedalTable
        list={selectAllCountries(listData)}
        onAddSilver={onAddSilver}
      />
    );

    wrapper.find('*[data-test="silver"]').simulate("click");

    expect(onAddSilver).toHaveBeenCalled();
    expect(onAddSilver).toHaveBeenCalledWith("UK", "add-silver");
  });

  it("should handle onAddBronze prop", () => {
    const wrapper = shallow(
      <MedalTable
        list={selectAllCountries(listData)}
        onAddBronze={onAddBronze}
      />
    );

    wrapper.find('*[data-test="bronze"]').simulate("click");

    expect(onAddBronze).toHaveBeenCalled();
    expect(onAddBronze).toHaveBeenCalledWith("UK", "add-bronze");
  });
});
