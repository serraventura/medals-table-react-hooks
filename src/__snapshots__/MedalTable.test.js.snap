// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MedalTable /> should match snapshot 1`] = `ReactWrapper {}`;

exports[`<MedalTable /> should match snapshot with data 1`] = `ReactWrapper {}`;
